pkg/operator/config.go

```diff
@@ -0,0 +1,163 @@
+	ClusterConfigName = "cluster-config-v1"
+	// InstallConfigKey is the key in the cluster config configmap containing yaml installConfig data
+	InstallConfigKey = "install-config"
+	// AWSPlatformType is used to install on AWS
+	AWSProvider = Provider("aws")
+	// LibvirtPlatformType is used to install of libvirt
+	LibvirtProvider = Provider("libvirt")
+	// OpenStackPlatformType is used to install on OpenStack
+	OpenStackProvider = Provider("openstack")
+)
+
+type Provider string
+
+// OperatorConfig contains configuration for MAO
+type OperatorConfig struct {
+	TargetNamespace string `json:"targetNamespace"`
+	Controller      string `json:"images"`
+}
+
+// Images allows build systems to inject images for MAO components
+type Images struct {
+	ClusterAPIControllerAWS       string `json:"clusterAPIControllerAWS"`
+	ClusterAPIControllerOpenStack string `json:"clusterAPIControllerOpenStack"`
+	ClusterAPIControllerLibvirt   string `json:"clusterAPIControllerLibvirt"`
+}
+
+// InstallConfig contains the mao relevant config coming from the install config, i.e provider
+type InstallConfig struct {
+	IntallPlatform `json:"platform"`
+}
+
+// IntallPlatform is the configuration for the specific platform upon which to perform
+// the installation. Only one of the platform configuration should be set
+type IntallPlatform struct {
+	// AWS is the configuration used when running on AWS
+	AWS interface{} `json:"aws,omitempty"`
+
+	// Libvirt is the configuration used when running on libvirt
+	Libvirt interface{} `json:"libvirt,omitempty"`
+
+	// OpenStack is the configuration used when running on OpenStack
+	OpenStack interface{} `json:"openstack,omitempty"`
+}
+
+func getInstallConfig(client kubernetes.Interface) (*InstallConfig, error) {
+	cm, err := client.CoreV1().ConfigMaps(ClusterConfigNamespace).Get(ClusterConfigName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed getting clusterconfig %s/%s: %v", ClusterConfigNamespace, ClusterConfigName, err)
+	}
+
+	return getInstallConfigFromClusterConfig(cm)
+}
+
+// getInstallConfigFromClusterConfig builds an install config from the cluster config.
+func getInstallConfigFromClusterConfig(clusterConfig *corev1.ConfigMap) (*InstallConfig, error) {
+	icYaml, ok := clusterConfig.Data[InstallConfigKey]
+	if !ok {
+		return nil, fmt.Errorf("missing %q in configmap", InstallConfigKey)
+	}
+	var ic InstallConfig
+	if err := yaml.Unmarshal([]byte(icYaml), &ic); err != nil {
+		return nil, fmt.Errorf("invalid InstallConfig: %v\nyaml:\n%s", err, icYaml)
+	}
+	return &ic, nil
+}
+
+func getProviderFromInstallConfig(installConfig *InstallConfig) (Provider, error) {
+	v := reflect.ValueOf(installConfig.IntallPlatform)
+	var nonNilFields int
+
+	for i := 0; i < v.NumField(); i++ {
+		if v.Field(i).Interface() != nil {
+			nonNilFields = nonNilFields + 1
+		}
+		if nonNilFields > 1 {
+			return "", fmt.Errorf("more than one platform provider given")
+		}
+	}
+
+	if installConfig.AWS != nil {
```

Comment:
This way AWS > Libvirt > OpenStack. What happens in case one specifies both AWS and Libvirt? Maybe add a check there can be only one provider set?

Reply:
that's checked here https://github.com/openshift/machine-api-operator/pull/106/files#diff-614eabe934833da7d7ec9938dfb3f1faR97 adn unit tested here https://github.com/openshift/machine-api-operator/pull/106/files#diff-b77facdf2f36a2cc873ec9ea3fae6b00R108